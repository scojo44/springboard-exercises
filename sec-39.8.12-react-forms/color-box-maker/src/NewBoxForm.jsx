import React, {useState} from "react";

const EMPTY_FORM = {
  color: '',
  width: '',
  height: ''
};

const NewBoxForm = ({addBox}) => {
  function handleSubmit(e) {
    e.preventDefault();
    addBox(formData);
    setFormData(() => {
      return {...EMPTY_FORM};
    });
  }

  function handleChange(e) {
    const {name, value} = e.target;

    setFormData(data => {
      return {
        ...data,
        [name]: value
      };
    })
  }

  const [formData, setFormData] = useState(EMPTY_FORM);

  return (
    <form onSubmit={handleSubmit} className="NewBoxForm">
      <p>
        <label htmlFor="color">Color: </label>
        <input type="text" id="color" name="color" value={formData.color} onChange={handleChange} />
      </p>
      <p>
        <label htmlFor="width">Width: </label>
        <input type="text" id="width" name="width" value={formData.width} onChange={handleChange} />
      </p>
      <p>
        <label htmlFor="height">Height: </label>
        <input type="text" id="height" name="height" value={formData.height} onChange={handleChange} />
      </p>
      <button type="submit">Add Box</button>
    </form>
  )
};

export default NewBoxForm;
