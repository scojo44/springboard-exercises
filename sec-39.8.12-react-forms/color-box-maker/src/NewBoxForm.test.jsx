import { render, fireEvent, queryByLabelText } from '@testing-library/react';
import NewBoxForm from './NewBoxForm';

describe('NewBoxForm Tests', () => {
  it('renders without crashing', () => {
    render(<NewBoxForm addBox={x => true} />);
  });

  it('matches snapshot', () => {
    const {asFragment} = render(<NewBoxForm addBox={x => true} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('shows the form', () => {
    const {getByLabelText} = render(<NewBoxForm addBox={x => true} />);
    const color = getByLabelText('Color:');
    const width = getByLabelText('Width:');
    const height = getByLabelText('Height:');
    expect(color).toBeInTheDocument();
    expect(width).toBeInTheDocument();
    expect(height).toBeInTheDocument();
  });

  it('submits the form, and calls the parent function with the formData', () => {
    let newBox = null;
    const addBox = vitest.fn(formData => {
      newBox = formData;
    });
    const {getByText, getByLabelText} = render(<NewBoxForm addBox={addBox} />);
    const color = getByLabelText('Color:');
    const width = getByLabelText('Width:');
    const height = getByLabelText('Height:');
    const button = getByText('Add Box');
    fireEvent.change(color, {target: {value: 'pink'}});
    fireEvent.change(width, {target: {value: '1em'}});
    fireEvent.change(height, {target: {value: '1em'}});
    fireEvent.click(button);
    expect(addBox).toHaveBeenCalled();
    expect(newBox).toEqual({
      color: 'pink',
      width: '1em',
      height: '1em'
    });
  });
});
