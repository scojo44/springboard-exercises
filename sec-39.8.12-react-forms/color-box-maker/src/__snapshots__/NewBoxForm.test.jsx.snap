// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NewBoxForm Tests > matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="NewBoxForm"
  >
    <p>
      <label
        for="color"
      >
        Color: 
      </label>
      <input
        id="color"
        name="color"
        type="text"
        value=""
      />
    </p>
    <p>
      <label
        for="width"
      >
        Width: 
      </label>
      <input
        id="width"
        name="width"
        type="text"
        value=""
      />
    </p>
    <p>
      <label
        for="height"
      >
        Height: 
      </label>
      <input
        id="height"
        name="height"
        type="text"
        value=""
      />
    </p>
    <button
      type="submit"
    >
      Add Box
    </button>
  </form>
</DocumentFragment>
`;
